fix nested external

Tom-van-Woudenberg opened the pull request:
# Fix: Recursive Conversion of Nested `external` Keys in ToC

## Summary

This pull request addresses a bug where only the first occurrence of an `external` key in the Table of Contents (ToC) YAML was converted to a local file reference, while any nested `external` keys within child fields (such as sections of a chapter) remained unconverted. Now, all `external` entries—no matter how deeply nested—are correctly converted to local paths.

## Root Cause

Previously, the function responsible for converting `external` keys (`external_to_local`) was only applied to the first level where an `external` key appeared. When an `external` entry was found, it would be replaced with a local `file` reference, but any child fields (like `sections`) within that same mapping were not recursively checked for additional `external` keys. This meant that nested `external` entries would remain untouched, leading to incomplete conversion in more complex ToCs.

## Solution

- Updated `external_to_local` to recursively process all child fields (such as `sections`, `chapters`, etc.) within a mapping after converting an `external` key.
- This ensures that after an `external` key is replaced with a `file` reference, any further `external` keys found deeper in the structure are also properly converted.

## Example

Given the following ToC snippet:

```yaml
- file: ODE/lesson.md
  sections:
  - external: https://github.com/TUDelft-books/CT1000/blob/CTB2210/book/week_2/session_1/intro.ipynb
    sections:
    - external: https://github.com/TUDelft-books/CEG-mechanics-BSc/blob/NL/book/tools/TI-84.md
```

Previously, only the first `external` would be converted. After this fix, both the outer and nested `external` fields will be converted to local `file` references.

## Impact

- Fully supports nested external content references in the ToC.
- No breaking changes to the ToC format or other features.

Tom-van-Woudenberg committed:
Update process_toc.py

Tom-van-Woudenberg committed:
Ruff lint error fix

Tom-van-Woudenberg committed:
Actually did ruff locally

Tom-van-Woudenberg commented:
@BSchilperoort , done! With some help from a friend :)

BSchilperoort commented:
Nice work, sorry for the late review.

I only have one remark, can you adjust the `_toc.yml` file in the external content test book? This would ensure this feature runs OK. I.e.;

```yml
    - external: https://gitlab.tudelft.nl/interactivetextbooks-citg/risk-and-reliability/-/blob/v0.1/book/intro.md
      sections:
      - external: https://gitlab.tudelft.nl/interactivetextbooks-citg/risk-and-reliability/-/blob/main/book/intro.md
        sections:
        - external: https://github.com/EXCITED-CO2/workshop_tutorial/blob/v1.0.0/book/extra_file.rst
```

Tom-van-Woudenberg committed:
Update _toc.yml

As demanded here: https://github.com/TeachBooks/TeachBooks/pull/108#issuecomment-3073694926

Tom-van-Woudenberg commented:
> I only have one remark, can you adjust the `_toc.yml` file in the external content test book? This would ensure this feature runs OK. 

I think I did what you @BSchilperoort  asked for here: https://github.com/TeachBooks/TeachBooks/pull/108/commits/3e16f135341bb4aab4856e3129fefa9f39de4cf7

Do you merge and release etc.?

BSchilperoort reviewed (approved):
Thanks for figuring this out, and updating the test file!
